### PR TITLE
test: Implement Firefox testing with CDP

### PR DESCRIPTION
**Pull request by marusak**
Firefox CDP is only partially implemented and some major features are missing. Also `chrome-remote-interface` misses some features on Firefox. Therefore I've created new cdp-driver which contains some hacks and workarounds. Since it uses very small subset of cdp protocol, it could be used for chrome as well, but lets keep also the proper driver and hopefully one day we can use that one for all browsers.

How to run this: (see note below)
prerequisite: Firefox browser (tested with current F31 Firefox build `69.0.1-3`)

`TEST_BROWSER=firefox ./test/verify/check-pages TestPages.testBasic -vt`

TODO:
- [ ] Implement screenshots
- [ ] opening new tabs in Firefox (not needed for CI or for TEST_WITH_UI, but when connecting to running browser)
- [x] testBasic (check_dashboard.TestBasicDashboard)
- [x] testNoConnectionSettings (check_networking_settings.TestNetworking)

[1](https://209.132.184.41:8493/logs/pull-12971-20191029-135353-b0973c96-cockpit-project-cockpit--fedora-30-firefox/log.html#105-2) [2](https://209.132.184.41:8493/logs/pull-12971-20191105-091919-295c0d2c-cockpit-project-cockpit--fedora-30-firefox/log.html)

### CODE REVIEW

**Comment by marusak** [comment]
New Firefox (firefox-70.0-1.fc31.x86_64 and current nightly) breaks propagation of new contexts. I cannot think of any workaround for this since this is a core functionality that we need. I reported it [here](https://bugzilla.mozilla.org/show_bug.cgi?id=1593226). I can either enforce firefox <= 69 in CI or block on that.

**Comment by marusak** [comment]
uuu, fancy! Only one failure, but it is currently broken `testLibvirt` one and one test was retried. I think this is as good as it gets :)

**Comment by martinpitt** [comment]
#13113 landed, this needs a rebase now.

File: test/common/cdp.py [hunk 1]
```diff
@@ -137,10 +159,45 @@ def find_cdp_port(self):
...
+
+        if self.browser == "chromium":
+            return [exe, "--headless", "--disable-gpu", "--no-sandbox", "--disable-setuid-sandbox",
+                    "--disable-namespace-sandbox", "--disable-seccomp-filter-sandbox",
+                    "--disable-sandbox-denial-logging", "--disable-pushstate-throttle",
+                    "--window-size=1920x1200", "--remote-debugging-port=%i" % cdp_port, "about:blank"]
+        elif self.browser == "firefox":
+            subprocess.Popen(["firefox", "--headless", "--no-remote", "-CreateProfile", "blank"], env=env).communicate()
+            p = subprocess.Popen("ls ~/.mozilla/firefox", shell=True, stdout=subprocess.PIPE, env=env)
+            in_profile = p.communicate()[0].decode("utf-8").split("\n")
```
Comment (martinpitt) [inline test/common/cdp.py]: Please use [os.listdir()](https://docs.python.org/3/library/os.html?highlight=listdir#os.listdir) here. Also, writing `~` here makes me terribly nervous -- we set `$HOME` here, but let's not use `os.path.expanduser()` here or ~, but explicitly `sef._browser_home` to make sure that we don't write into the real user home.

File: test/common/cdp.py [hunk 2]
```diff
@@ -137,10 +159,45 @@ def find_cdp_port(self):
...
+            userjs = "~/.mozilla/firefox/"
+            handler = "~/.mozilla/firefox/"
+            for item in in_profile:
+                if item.endswith("blank"):
+                    handler = os.path.join(handler, item, "handlers.json")
+                    userjs = os.path.join(userjs, item, "user.js")
+                    break
```
Comment (martinpitt) [inline test/common/cdp.py]: Looks like you don't really need to read the whole directory in the first place? How is that different to simply doing `os.path.join(sef._browser_home, "blank")`? (And then tacking on the two *.json, of course).

File: test/common/cdp.py [hunk 3]
```diff
@@ -137,10 +159,45 @@ def find_cdp_port(self):
...
+            user_pref("datareporting.policy.dataSubmissionEnabled", false);
+            user_pref("toolkit.telemetry.reportingpolicy.firstRun", false);
+            user_pref("dom.disable_beforeunload", true);
+            user_pref("browser.download.dir", "{0}");
+            user_pref("browser.download.folderList", 2);
+            """.format(self.download_dir)
+
+            subprocess.Popen("echo '{0}' > {1}".format(prefs, userjs), shell=True, env=env).communicate()
+            subprocess.Popen("echo '{\"defaultHandlersVersion\":{\"en-US\":4},\"mimeTypes\":{\"application/xz\":{\"action\":0,\"extensions\":[\"xz\"]}}}' > " + handler,
+                             shell=True, env=env).communicate()
```
Comment (martinpitt) [inline test/common/cdp.py]: It's easier and much more robust to just do this in Python. Something like
```py
with open(os.path.join(sef._browser_home, "blank", ""user.js"), "w") as f:
    f.write("blabla")
```

For handlers.json you can pick whether you want to use `json.dump(object, f)` or write the JSON manually (not much difference for such small objects)

File: test/common/cdp.py [hunk 4]
```diff
@@ -209,6 +262,10 @@ def start(self):
                 src = f.read()
             # HACK: injecting sizzle fails on missing `document` in assert()
             src = src.replace('function assert( fn ) {', 'function assert( fn ) { return true;')
+            # HACK: sizzle tracks document and when we switch frames, it sees the old document
+            # although we execute it in different context.
```
Comment (martinpitt) [inline test/common/cdp.py]: Meh, about time we finally kick out sizzle.. Most of our selectors should be just fine with standard CSS selectors now, and we can add some helpers to test-functions.js for the one or two remaining use cases that still need sizzle. `</thinking_aloud>`

File: test/common/testlib.py [hunk 5]
```diff
@@ -571,14 +600,21 @@ def snapshot(self, title, label=None):
...
             filename = "{0}-{1}.png".format(label or self.label, title)
-            ret = self.cdp.invoke("Page.captureScreenshot", no_trace=True)
-            if "data" in ret:
-                with open(filename, 'wb') as f:
-                    f.write(base64.standard_b64decode(ret["data"]))
-                attach(filename)
-                print("Wrote screenshot to " + filename)
-            else:
-                print("Screenshot not available")
+            if (self.cdp.browser == "chromium"):
```
Comment (martinpitt) [inline test/common/testlib.py]: No parens around if or  "elif" below, please.

File: test/common/testlib.py [hunk 6]
```diff
@@ -571,14 +600,21 @@ def snapshot(self, title, label=None):
...
+                ret = self.cdp.invoke("Page.captureScreenshot", no_trace=True)
```
Comment (martinpitt) [inline test/common/testlib.py]: Is this API just not implemented, or buggy? In the latter case, is there a bug report for that?

File: test/common/testlib.py [hunk 7]
```diff
@@ -571,14 +600,21 @@ def snapshot(self, title, label=None):
...
+                    with open(filename, 'wb') as f:
+                        f.write(base64.standard_b64decode(ret["data"]))
+                    attach(filename)
+                    print("Wrote screenshot to " + filename)
+                else:
+                    print("Screenshot not available")
+            elif (self.cdp.browser == "firefox"):
+                # TODO: implement me
+                # Something like:
+                # Runtime.execute(':screenshot --file <path>)
```
Comment (martinpitt) [inline test/common/testlib.py]: Is there some API like that already, and thus this is a SMOP? Or completely lacking right now? (I would think that for puppeteer support firefox must have some screenshot support?

File: test/verify/check-sosreport [hunk 8]
```diff
@@ -52,14 +50,15 @@ threads=2
             b.wait(lambda: b.eval_js('ph_find(".progress-bar").offsetWidth') > 0)
             b.wait_visible("#sos-download")
 
-        download_dir = tempfile.mkdtemp()
-        self.addCleanup(shutil.rmtree, download_dir)
-        b.cdp.invoke("Page.setDownloadBehavior", behavior="allow", downloadPath=download_dir)
+        for f in glob.glob(os.path.join(b.cdp.download_dir, '*')):
```
Comment (martinpitt) [inline test/verify/check-sosreport]: OK, but more robust to do that with `os.listdir()`. But why does that need to be cleaned in the first place? Every test case starts with a fresh Browser and thus CDP object, and thus a new empty download dir?

File: test/verify/check-menu [hunk 9]
```diff
@@ -44,7 +44,10 @@ class TestMenu(MachineCase):
         b.click('#about button[data-dismiss="modal"]')
         b.wait_popdown("about")
 
-        self.check_axe("Test-navigation")
+        # HACK: We cannot test axe on Firefox since `axe.run()` returns promise
+        # and Firefox CDP cannot wait for it to resolve
+        if b.cdp.browser != "firefox":
+            self.check_axe("Test-navigation")
```
Comment (martinpitt) [inline test/verify/check-menu]: Please put this hack into the `check_axe()` definition itself, instead of into every call (well, two right now, but we'll get more)

File: test/common/firefox-cdp-driver.js [hunk 10]
```diff
@@ -0,0 +1,425 @@
...
+}
+
+/**
+ * Format response to the client
+ */
+
+function fatal() {
+    console.error.apply(console.error, arguments);
+    process.exit(1);
+}
```
Comment (martinpitt) [inline test/common/firefox-cdp-driver.js]: There's a lot of duplicated code here. What's the next step for these two files? Do you want to work on unifying them again, or are there enough differences that these need to be two separate drivers for a while? In the latter case, could these move out the commonalities into a library?

I only coarsely reviewed this file, as it's mostly a question of "does it work?" and that smells like it's going to get more cleanup in the near future. :grin:

File: test/common/cdp.py (hunk 2 above)
Comment (marusak) [reply test/common/cdp.py]: The path looks something like `~/.mozilla/firefox/ldsyl7au.blank/`. I could use `glob` though?

File: test/common/cdp.py (hunk 4 above)
Comment (marusak) [reply test/common/cdp.py]: yeah, I tried to kick out sizzle, see https://github.com/cockpit-project/cockpit/pull/13015

File: test/common/testlib.py (hunk 6 above)
Comment (marusak) [reply test/common/testlib.py]: Not implemented. There is tracker anyway https://bugzilla.mozilla.org/show_bug.cgi?id=1549466

File: test/common/testlib.py (hunk 7 above)
Comment (marusak) [reply test/common/testlib.py]: The API (`Page.captureScreenshot`) is missing. You can execute in your firefox console `:screenshot` and it will create screenshot. I though we could do this and use `Runtime.execute` as it is what we do normally and it behaves the same as when you execute something in console. I tried it but it did not work. I haven't yet dug deep why it is not working.

File: test/common/cdp.py (hunk 2 above)
Comment (martinpitt) [reply test/common/cdp.py]: Argh, sorry -- I misread the code. Yes, glob seems fine indeed.

File: test/common/testlib.py (hunk 7 above)
Comment (martinpitt) [reply test/common/testlib.py]: Ack, no problem. Thus for now, debugging a failing test would need to go through attaching it to an interactive firefox.

File: test/verify/check-sosreport (hunk 8 above)
Comment (marusak) [reply test/verify/check-sosreport]: Right, this was thinko since it is created in CDP init and this it is clean for every test.

File: test/common/firefox-cdp-driver.js (hunk 10 above)
Comment (marusak) [reply test/common/firefox-cdp-driver.js]: Well, in a ideal world we would use only one file (the chromium one) and I hope that in some time we can do that. However right now this firefox driver contains a lot of hacks and workaround since the firefox API is very limited.
I don't think we should invest much time to unify them now, but as firefox implements more and more API I'll be happy to  move them closer to each other.  I would also like to point out that firefox CDP is becoming active and things are moving forward so I really hope that the functionality we need could be implemented in some reasonable time frame.

> could these move out the commonalities into a library?

I guess? There are some common functions but they all interact with `process` - which I am not sure if it can be provided by library, but I guess why not? I mostly just did not want to touch the chromium driver. But if you think it is worth it I'll go for it

File: test/common/cdp.py [hunk 11]
```diff
@@ -74,7 +94,9 @@ def invoke(self, fn, **kwargs):
         # frame support for Runtime.evaluate(): map frame name to
         # executionContextId and insert into argument object; this must not be quoted
         # see "Frame tracking" in cdp-driver.js for how this works
-        if fn == 'Runtime.evaluate' and self.cur_frame:
+        # Also Firefox always needs context, cannot work with default context.
+        # TODO: it should be fairly simple for chromium to always work with contexts
```
[the remark this answers is not in the document]
Comment (marusak) [reply test/common/cdp.py]: I just did small edit in chrome driver and now this was dropped

File: test/common/firefox-cdp-driver.js (hunk 10 above)
Comment (martinpitt) [reply test/common/firefox-cdp-driver.js]: Not right now. The chromium driver hasn't changed in some time. We should do that if/when we need to change it, but chances are that some firefox hacks went away until then. So indeed it seems sensible for now to wait.

File: test/common/cdp.py [hunk 12]
```diff
@@ -175,27 +173,21 @@ def browser_cmd(self, cdp_port, env):
...
+                    user_pref("remote.enabled", true);
+                    user_pref("datareporting.policy.dataSubmissionEnabled", false);
+                    user_pref("toolkit.telemetry.reportingpolicy.firstRun", false);
+                    user_pref("dom.disable_beforeunload", true);
+                    user_pref("browser.download.dir", "{0}");
+                    user_pref("browser.download.folderList", 2);
+                    """.format(self.download_dir))
+
+            with open(os.path.join(profile, "handlers.json"), "w") as f:
+                f.write('{\"defaultHandlersVersion\":{\"en-US\":4},\"mimeTypes\":{\"application/xz\":{\"action\":0,\"extensions\":[\"xz\"]}}}')
```
Comment (martinpitt) [inline test/common/cdp.py]: This doesn't need the `\`.

**Comment by marusak** [comment]
@martinpitt I had to repush as after rebase one test needed a small tweak. Can you please reapprove?

**Comment by martinpitt** [comment]
[testPrivateKeys](https://209.132.184.41:8493/logs/pull-12971-20191111-195252-5ac0eb76-cockpit-project-cockpit--ubuntu-1804/log.html) fails an awful lot here. That doesn't ring a bell, I rather retry that one.

**Comment by marusak** [comment]
`testPrivateKeys` again failed on ubuntu-1804 but not anywhere else. 
But it also happens in different PRs like [here](https://209.132.184.41:8493/logs/pull-13125-20191111-134646-087341bf-cockpit-project-cockpit--ubuntu-1804/log.html).
We don't need to block this on it. I'll pull this image and debug it locally.

**Comment by marusak** [comment]
Fixed in #13131
